components/webext-storage/src/api.rs

```diff
@@ -0,0 +1,425 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use rusqlite::Connection;
+use serde_derive::Serialize;
+use serde_json::{Map, Value as JsonValue};
+use sql_support::{self, ConnExt};
+use std::collections::HashMap;
+
+const QUOTA_BYTES: usize = 102_400;
+const QUOTA_BYTES_PER_ITEM: usize = 8_192;
+const MAX_ITEMS: usize = 512;
+// Note there are constants for "operations per minute" etc, which aren't
+// enforced here.
+
+type JsonMap = Map<String, JsonValue>;
+
+fn get_from_db(conn: &Connection, ext_id: &str) -> Result<Option<JsonMap>> {
+    Ok(
+        match conn.try_query_one::<String>(
+            "SELECT data FROM moz_extension_data
+             WHERE ext_id = :ext_id",
+            &[(":ext_id", &ext_id)],
+            true,
+        )? {
+            Some(s) => match serde_json::from_str(&s)? {
+                JsonValue::Object(m) => Some(m),
+                // we could panic here as it's theoretically impossible, but we
+                // might as well treat it as not existing...
+                _ => None,
+            },
+            None => None,
+        },
+    )
+}
+
+fn save_to_db(conn: &Connection, ext_id: &str, val: &JsonValue) -> Result<()> {
+    // Convert to bytes so we can enforce the quota.
+    let sval = val.to_string();
+    let bytes: Vec<u8> = sval.bytes().collect();
+    if bytes.len() > QUOTA_BYTES {
+        return Err(ErrorKind::QuotaError(QuotaReason::TotalBytes).into());
+    }
+    // XXX - work out how to get use these bytes directly instead of sval, so
+    // we don't utf-8 encode twice!
+
+    // XXX - sync support will need to do the syncStatus thing here.
+    conn.execute_named(
+        "INSERT OR REPLACE INTO moz_extension_data(ext_id, data)
+            VALUES (:ext_id, :data)",
+        &[(":ext_id", &ext_id), (":data", &sval)],
+    )?;
+    Ok(())
+}
+
+fn remove_from_db(conn: &Connection, ext_id: &str) -> Result<()> {
+    // XXX - sync support will need to do the tombstone thing here.
+    conn.execute_named(
+        "DELETE FROM moz_extension_data
+        WHERE ext_id = :ext_id",
+        &[(":ext_id", &ext_id)],
+    )?;
+    Ok(())
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize)]
```

Comment:
Any chance of us accidentally logging PII if we use the default `Debug` impl in a log message?

Reply:
Yes - but it seems quite a burden to set the bar at "can't derive Debug if what it holds might be pii" - and that's far from true today and would make debugging/diagnosing tests really problematic - many of these structs get Debug added when I find myself trying to println! them :) That said though, a low-cost way of helping to enforce what we desire is worth investigating.